### PR TITLE
Remove unused state_id foreign key

### PR DESCRIPTION
**Pull request by jlowin**
**Thanks for contributing to Prefect Server!**

Please describe your work and make sure your PR:

- [x] adds new tests (if appropriate)
- [x] adds a changelog entry to the `changes/` directory (if appropriate)

Note that your PR will not be reviewed unless these two boxes are checked.

## What does this PR change?
This PR removes a foreign key on `task_run.state_id` and `flow_run.state_id` that once pointed to the current state; it is unused and creates performance issues.

### CODE REVIEW

File: services/postgres/alembic/env.py [hunk 1]
```diff
@@ -67,12 +67,12 @@ def apply_hasura_metadata(context):
...
     if fn and fn.__name__ == "upgrade":
-        if context.get_context().opts.get("destination_rev") == "heads":
+        if script.revision == context.script.get_current_head():
             hasura_revision = None
         else:
-            hasura_revision = script.down_revision
+            hasura_revision = script.up_revision
     elif fn and fn.__name__ == "downgrade":
-        hasura_revision = script.revision
+        hasura_revision = script.down_revision
```
Comment (cicdw) [inline services/postgres/alembic/env.py:75]: why are you having to touch this file?

File: services/postgres/alembic/versions/2020-08-08T205817_remove_state_id_foreign_keys.py [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+        DROP FUNCTION update_task_run_with_latest_state CASCADE;
+        """
+    )
+    op.execute(
+        """
+        CREATE FUNCTION update_flow_run_with_latest_state() RETURNS trigger
+            LANGUAGE plpgsql
+            AS $$
+                    BEGIN
+                        UPDATE flow_run
```
Comment (cicdw) [inline services/postgres/alembic/versions/2020-08-08T205817_remove_state_id_foreign_keys.py:34]: is this trigger going to run every time a task run sends a heartbeat?

File: services/postgres/alembic/versions/2020-08-08T205817_remove_state_id_foreign_keys.py [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+                                task_run.version IS NULL
+                                OR (
+                                    new_latest_state.version >= COALESCE(task_run.version, -1)
+                                    AND new_latest_state.timestamp > COALESCE(task_run.state_timestamp, '2000-01-01')));
+                        RETURN NEW;
+                    END;
+                    $$;
+
+            CREATE TRIGGER update_flow_run_after_inserting_state
+            AFTER INSERT ON flow_run_state
```
Comment (cicdw) [inline services/postgres/alembic/versions/2020-08-08T205817_remove_state_id_foreign_keys.py:114]: Ah - if I'm reading this correctly the trigger only runs after a new state is placed in this table, yea?

Comment (jlowin) [reply services/postgres/alembic/versions/2020-08-08T205817_remove_state_id_foreign_keys.py:114]: That's correct, this is the trigger that populates the `flow_run` table with the info from the latest flow run state. It has to be modified here to no longer update the `state_id`

File: services/postgres/alembic/env.py (hunk 1 above)
Comment (jlowin) [reply services/postgres/alembic/env.py:75]: There's a small bug in this code (two actually) which is exposed by destructive migrations like this one -- essentially that it only runs properly if you pass the string `"heads"` as the alembic target. Our unit tests for migrations don't pass `"heads"`; they pass an explicit number of migrations to run (`"+2"`). Those didn't trip this logic correctly and attempted to apply a hasura metadata file that was not compatible with the database state.

**Pull request merged**